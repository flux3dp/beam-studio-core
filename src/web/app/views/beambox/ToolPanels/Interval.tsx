import classNames from 'classnames';
import React from 'react';

import BeamboxPreference from 'app/actions/beambox/beambox-preference';
import Constant from 'app/actions/beambox/constant';
import i18n from 'helpers/i18n';
import storage from 'implementations/storage';
import UnitInput from 'app/widgets/Unit-Input-v2';

const LANG = i18n.lang.beambox.tool_panels;

interface Props {
  dx?: number,
  dy?: number,
  onValueChange?: (rc: { dx: number, dy: number }) => void,
}

interface State {
  dx: number,
  dy: number,
  isCollapsed: boolean,
}

class Interval extends React.Component<Props, State> {
  constructor(props) {
    super(props);
    const { dx, dy } = this.props;
    this.state = {
      dx,
      dy,
      isCollapsed: false,
    };
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    this.setState({
      dx: nextProps.dx,
      dy: nextProps.dy,
    });
  }

  onDxChanged = (dx: number) => {
    const { onValueChange } = this.props;
    const { dy } = this.state;
    onValueChange({
      dx,
      dy,
    });
    this.setState({ dx });
  };

  onDyChanged = (dy: number) => {
    const { onValueChange } = this.props;
    const { dx } = this.state;
    onValueChange({
      dx,
      dy,
    });
    this.setState({ dy });
  };

  getValueCaption = (): string => {
    const { dx, dy } = this.state;
    const units = storage.get('default-units') || 'mm';
    return units === 'inches'
      ? `${Number(dx / 25.4).toFixed(3)}\", ${Number(dy / 25.4).toFixed(3)}\"`
      : `${dx}, ${dy} mm`;
  };

  render(): JSX.Element {
    const { dx, dy, isCollapsed } = this.state;
    return (
      <div className="tool-panel">
        <label className="controls accordion">
          <input type="checkbox" className="accordion-switcher" defaultChecked />
          <p className="caption" onClick={() => this.setState({ isCollapsed: !isCollapsed })}>
            {LANG.array_interval}
            <span className="value">{this.getValueCaption()}</span>
          </p>
          <div className={classNames('tool-panel-body', { collapsed: isCollapsed })}>
            <div className="control">
              <span className="text-center header">{LANG.dx}</span>
              <UnitInput
                min={0}
                max={Constant.dimension.getWidth(BeamboxPreference.read('workarea')) / Constant.dpmm}
                unit="mm"
                defaultValue={dx}
                getValue={this.onDxChanged}
              />
            </div>
            <div className="control">
              <span className="text-center header">{LANG.dy}</span>
              <UnitInput
                min={0}
                max={Constant.dimension.getHeight(BeamboxPreference.read('workarea')) / Constant.dpmm}
                unit="mm"
                defaultValue={dy}
                getValue={this.onDyChanged}
              />
            </div>
          </div>
        </label>
      </div>
    );
  }
}

export default Interval;
