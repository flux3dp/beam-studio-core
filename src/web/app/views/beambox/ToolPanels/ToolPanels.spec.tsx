/* eslint-disable import/first */
import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

jest.mock('helpers/i18n', () => ({
  lang: {
    beambox: {
      tool_panels: {
        cancel: 'Cancel',
        confirm: 'Confirm',
        grid_array: 'Create Grid Array',
        offset: 'Offset',
      },
    },
  },
}));

const get = jest.fn();
jest.mock('implementations/storage', () => ({
  get,
}));

const getSVGAsync = jest.fn();
jest.mock('helpers/svg-editor-helper', () => ({
  getSVGAsync,
}));

const setMode = jest.fn();
const gridArraySelectedElement = jest.fn();
const setHasUnsavedChange = jest.fn();
const offsetElements = jest.fn();
const nestElements = jest.fn();
getSVGAsync.mockImplementation((callback) => {
  callback({
    Canvas: {
      setMode,
      gridArraySelectedElement,
      setHasUnsavedChange,
      offsetElements,
      nestElements,
    },
  });
});

jest.mock('app/views/beambox/ToolPanels/Interval', () => function DummyInterval() {
  return (<div>This is dummy Interval</div>);
});

jest.mock('app/views/beambox/ToolPanels/NestGAPanel', () => function DummyNestGAPanel() {
  return (<div>This is dummy NestGAPanel</div>);
});

jest.mock('app/views/beambox/ToolPanels/NestRotationPanel', () => function DummyNestRotationPanel() {
  return (<div>This is dummy NestRotationPanel</div>);
});

jest.mock('app/views/beambox/ToolPanels/NestSpacingPanel', () => function DummyNestSpacingPanel() {
  return (<div>This is dummy NestSpacingPanel</div>);
});

jest.mock('app/views/beambox/ToolPanels/OffsetCornerPanel', () => function DummyOffsetCornerPanel() {
  return (<div>This is dummy OffsetCornerPanel</div>);
});

jest.mock('app/views/beambox/ToolPanels/OffsetDirectionPanel', () => function DummyOffsetDirectionPanel() {
  return (<div>This is dummy OffsetDirectionPanel</div>);
});

jest.mock('app/views/beambox/ToolPanels/OffsetDistancePanel', () => function DummyOffsetDistancePanel() {
  return (<div>This is dummy OffsetDistancePanel</div>);
});

jest.mock('app/views/beambox/ToolPanels/RowColumn', () => function DummyRowColumn() {
  return (<div>This is dummy RowColumn</div>);
});

jest.mock('app/actions/beambox/toolPanelsController', () => ({
}));

import ToolPanels from './ToolPanels';

describe('should render correctly', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('type is gridArray', () => {
    const unmount = jest.fn();
    const wrapper = shallow(<ToolPanels
      type="gridArray"
      data={{
        rowcolumn: {
          row: 1,
          column: 1,
        },
        distance: {
          dx: 0,
          dy: 0,
        },
      }}
      unmount={unmount}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('type is offset', () => {
    const unmount = jest.fn();
    const wrapper = shallow(<ToolPanels
      type="offset"
      data={{
        rowcolumn: {
          row: 1,
          column: 1,
        },
        distance: {
          dx: 0,
          dy: 0,
        },
      }}
      unmount={unmount}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('type is nest', () => {
    const unmount = jest.fn();
    const wrapper = shallow(<ToolPanels
      type="nest"
      data={{
        rowcolumn: {
          row: 1,
          column: 1,
        },
        distance: {
          dx: 0,
          dy: 0,
        },
      }}
      unmount={unmount}
    />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
