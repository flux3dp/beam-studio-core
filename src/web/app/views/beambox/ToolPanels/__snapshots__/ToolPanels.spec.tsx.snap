// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly type is gridArray 1`] = `
<div
  className="tool-panels en"
  id="beamboxToolPanel"
  style={
    {
      "bottom": 10,
      "left": undefined,
      "position": "absolute",
      "zIndex": 999,
    }
  }
>
  <div
    className="tool-panel"
  >
    <label
      className="controls accordion"
    >
      <input
        className="accordion-switcher"
        defaultChecked={true}
        type="checkbox"
      />
      <p
        className="caption"
      >
        <span
          className="value"
        >
          Create Grid Array
        </span>
      </p>
    </label>
  </div>
  <DummyRowColumn
    column={1}
    key="rowColumn"
    onValueChange={[Function]}
    row={1}
  />
  <DummyInterval
    dx={0}
    dy={0}
    key="distance"
    onValueChange={[Function]}
  />
  <div
    className="tool-block"
  >
    <div
      className="btn-h-group"
    >
      <button
        className="btn btn-default primary"
        onClick={[Function]}
        type="button"
      >
        Confirm
      </button>
      <button
        className="btn btn-default"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`should render correctly type is nest 1`] = `
<div
  className="tool-panels en"
  id="beamboxToolPanel"
  style={
    {
      "bottom": 10,
      "left": undefined,
      "position": "absolute",
      "zIndex": 999,
    }
  }
>
  <div
    className="tool-panel"
  >
    <label
      className="controls accordion"
    >
      <input
        className="accordion-switcher"
        defaultChecked={true}
        type="checkbox"
      />
      <p
        className="caption"
      >
        <span
          className="value"
        />
      </p>
    </label>
  </div>
  <DummyNestSpacingPanel
    key="nestOffset"
    onValueChange={[Function]}
    spacing={0}
  />
  <DummyNestRotationPanel
    key="nestRotation"
    onValueChange={[Function]}
    rotations={1}
  />
  <DummyNestGAPanel
    key="nestGA"
    nestOptions={
      {
        "generations": 3,
        "population": 10,
        "rotations": 1,
        "spacing": 0,
      }
    }
    updateNestOptions={[Function]}
  />
  <div
    className="tool-block"
  >
    <div
      className="btn-h-group"
    >
      <button
        className="btn btn-default primary"
        onClick={[Function]}
        type="button"
      >
        Confirm
      </button>
      <button
        className="btn btn-default"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`should render correctly type is offset 1`] = `
<div
  className="tool-panels en"
  id="beamboxToolPanel"
  style={
    {
      "bottom": 10,
      "left": undefined,
      "position": "absolute",
      "zIndex": 999,
    }
  }
>
  <div
    className="tool-panel"
  >
    <label
      className="controls accordion"
    >
      <input
        className="accordion-switcher"
        defaultChecked={true}
        type="checkbox"
      />
      <p
        className="caption"
      >
        <span
          className="value"
        >
          Offset
        </span>
      </p>
    </label>
  </div>
  <DummyOffsetDirectionPanel
    dir={1}
    key="offsetDir"
    onValueChange={[Function]}
  />
  <DummyOffsetCornerPanel
    cornerType="sharp"
    key="offsetCorner"
    onValueChange={[Function]}
  />
  <DummyOffsetDistancePanel
    distance={5}
    key="offsetDist"
    onValueChange={[Function]}
  />
  <div
    className="tool-block"
  >
    <div
      className="btn-h-group"
    >
      <button
        className="btn btn-default primary"
        onClick={[Function]}
        type="button"
      >
        Confirm
      </button>
      <button
        className="btn btn-default"
        onClick={[Function]}
        type="button"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;
